File: ccwap/server/models/advanced.py
"""Pydantic models for advanced analytics dashboards."""

from typing import List, Optional, Dict, Any

from pydantic import BaseModel


class ReliabilitySummary(BaseModel):
    total_tool_calls: int = 0
    total_errors: int = 0
    error_rate: float = 0.0
    error_cost: float = 0.0


class ReliabilityHeatmapCell(BaseModel):
    tool_name: str
    error_category: str
    errors: int = 0
    error_cost: float = 0.0


class ParetoItem(BaseModel):
    label: str
    count: int = 0
    cost: float = 0.0


class FailingWorkflow(BaseModel):
    workflow: str
    from_tool: str
    to_tool: str
    branch: str
    failures: int = 0
    cost: float = 0.0


class ReliabilityResponse(BaseModel):
    summary: ReliabilitySummary
    heatmap: List[ReliabilityHeatmapCell] = []
    pareto_tools: List[ParetoItem] = []
    pareto_commands: List[ParetoItem] = []
    pareto_languages: List[ParetoItem] = []
    by_branch: List[Dict[str, Any]] = []
    top_failing_workflows: List[FailingWorkflow] = []


class BranchSummary(BaseModel):
    branch: str
    cost: float = 0.0
    errors: int = 0
    tool_calls: int = 0
    loc_written: int = 0
    cache_hit_rate: float = 0.0


class BranchTrendPoint(BaseModel):
    date: str
    branch: str
    cost: float = 0.0
    errors: int = 0
    tool_calls: int = 0
    loc_written: int = 0
    cache_hit_rate: float = 0.0


class BranchAnomaly(BaseModel):
    date: str
    branch: str
    cost: float = 0.0
    zscore: float = 0.0
    reason: str = "cost_spike"


class BranchHealthResponse(BaseModel):
    branches: List[BranchSummary] = []
    trend: List[BranchTrendPoint] = []
    anomalies: List[BranchAnomaly] = []


class PromptEfficiencySummary(BaseModel):
    total_sessions: int = 0
    sessions_with_thinking: int = 0
    sessions_with_truncation: int = 0
    high_cost_low_output_sessions: int = 0
    avg_cost_per_loc: float = 0.0


class PromptEfficiencyScatterPoint(BaseModel):
    session_id: str
    project: str
    model: str
    branch: str
    cost: float = 0.0
    thinking_chars: int = 0
    input_tokens: int = 0
    output_tokens: int = 0
    cache_read_tokens: int = 0
    loc_written: int = 0
    truncations: int = 0
    token_mix_ratio: float = 0.0
    output_per_cost: float = 0.0
    efficiency_score: float = 0.0


class PromptEfficiencyResponse(BaseModel):
    summary: PromptEfficiencySummary
    funnel: List[Dict[str, Any]] = []
    by_stop_reason: List[Dict[str, Any]] = []
    scatter: List[PromptEfficiencyScatterPoint] = []
    outliers: List[PromptEfficiencyScatterPoint] = []


class WorkflowTransition(BaseModel):
    from_tool: str
    to_tool: str
    count: int = 0
    failures: int = 0
    failure_rate: float = 0.0


class RetryLoop(BaseModel):
    session_id: str
    tool_name: str
    retries: int = 0
    branch: str
    user_type: str


class FailureHandoff(BaseModel):
    parent_session_id: str
    child_session_id: str
    branch: str
    handoff: str
    errors: int = 0
    error_rate: float = 0.0


class BlockedSession(BaseModel):
    session_id: str
    project: str
    branch: str
    user_type: str
    failures: int = 0
    retries: int = 0
    stall_score: int = 0


class WorkflowBottleneckResponse(BaseModel):
    transition_matrix: List[WorkflowTransition] = []
    retry_loops: List[RetryLoop] = []
    failure_handoffs: List[FailureHandoff] = []
    blocked_sessions: List[BlockedSession] = []

File: ccwap/frontend/src/api/advanced.ts
import { useQuery } from '@tanstack/react-query'
import { apiFetch, buildQuery } from './client'
import { advancedKeys } from './keys'
import type { DateRange } from '@/hooks/useDateRange'

export interface ReliabilityData {
  summary: {
    total_tool_calls: number
    total_errors: number
    error_rate: number
    error_cost: number
  }
  heatmap: Array<{ tool_name: string; error_category: string; errors: number; error_cost: number }>
  pareto_tools: Array<{ label: string; count: number; cost: number }>
  pareto_commands: Array<{ label: string; count: number; cost: number }>
  pareto_languages: Array<{ label: string; count: number; cost: number }>
  by_branch: Array<{ branch: string; errors: number; cost: number }>
  top_failing_workflows: Array<{ workflow: string; from_tool: string; to_tool: string; branch: string; failures: number; cost: number }>
}

export interface BranchHealthData {
  branches: Array<{ branch: string; cost: number; errors: number; tool_calls: number; loc_written: number; cache_hit_rate: number }>
  trend: Array<{ date: string; branch: string; cost: number; errors: number; tool_calls: number; loc_written: number; cache_hit_rate: number }>
  anomalies: Array<{ date: string; branch: string; cost: number; zscore: number; reason: string }>
}

export interface PromptEfficiencyPoint {
  session_id: string
  project: string
  model: string
  branch: string
  cost: number
  thinking_chars: number
  input_tokens: number
  output_tokens: number
  cache_read_tokens: number
  loc_written: number
  truncations: number
  token_mix_ratio: number
  output_per_cost: number
  efficiency_score: number
}

export interface PromptEfficiencyData {
  summary: {
    total_sessions: number
    sessions_with_thinking: number
    sessions_with_truncation: number
    high_cost_low_output_sessions: number
    avg_cost_per_loc: number
  }
  funnel: Array<{ stage: string; value: number }>
  by_stop_reason: Array<{ stop_reason: string; count: number; percentage: number }>
  scatter: PromptEfficiencyPoint[]
  outliers: PromptEfficiencyPoint[]
}

export interface WorkflowBottleneckData {
  transition_matrix: Array<{ from_tool: string; to_tool: string; count: number; failures: number; failure_rate: number }>
  retry_loops: Array<{ session_id: string; tool_name: string; retries: number; branch: string; user_type: string }>
  failure_handoffs: Array<{ parent_session_id: string; child_session_id: string; branch: string; handoff: string; errors: number; error_rate: number }>
  blocked_sessions: Array<{ session_id: string; project: string; branch: string; user_type: string; failures: number; retries: number; stall_score: number }>
}

export function useReliability(dateRange: DateRange) {
  return useQuery({
    queryKey: advancedKeys.reliability(dateRange),
    queryFn: () => apiFetch<ReliabilityData>(`/reliability${buildQuery({ from: dateRange.from, to: dateRange.to })}`),
  })
}

export function useBranchHealth(dateRange: DateRange, branches?: string | null) {
  return useQuery({
    queryKey: advancedKeys.branchHealth(dateRange, branches),
    queryFn: () => apiFetch<BranchHealthData>(
      `/branch-health${buildQuery({ from: dateRange.from, to: dateRange.to, branches })}`
    ),
  })
}

export function usePromptEfficiency(dateRange: DateRange) {
  return useQuery({
    queryKey: advancedKeys.promptEfficiency(dateRange),
    queryFn: () => apiFetch<PromptEfficiencyData>(
      `/prompt-efficiency${buildQuery({ from: dateRange.from, to: dateRange.to })}`
    ),
  })
}

export function useWorkflowBottlenecks(dateRange: DateRange) {
  return useQuery({
    queryKey: advancedKeys.workflowBottlenecks(dateRange),
    queryFn: () => apiFetch<WorkflowBottleneckData>(
      `/workflow-bottlenecks${buildQuery({ from: dateRange.from, to: dateRange.to })}`
    ),
  })
}
